UNITTESTS/README.md

```diff
@@ -0,0 +1,288 @@
+```
+UNITTESTS
+  ├── mbed_unittest.py                       Script for building and running unit tests
+  ├── CMakeLists.txt                         CMake project definition file
+  ├── CMakeSettings.json                     CMake configurations for Visual Studio 2017
+  ├── README.md
+  ├── googletest-CMakeLists.txt.in           CMake project definition file for Google Test
+  │
+  ├── features
+  │   └── netsocket                          Directory tree that mirrors Mbed OS root
+  │       ├── NetworkInterface               Name of the class to be tested
+  │       │   ├── test_NetworkInterface.cpp
+  │       │   └── unittest.cmake             CMake module for unit test
+  │       └── Socket
+  │
+  ├── stubs                                  Stubs used in tests.
+  ├── target_h                               Stub headers used in tests.
+  ├── Modules                                CMake modules
+  └── template                               Templates for quickly generating unittest tree
+```
+Each unit test has an identical directory tree as seen in the Mbed OS root folder. This is not a mandatory requirement but helps to maintain tests. Each class to be tested have their own `unittest.cmake` which is found by `CMakeLists.txt`.
+
+## Creating a unit test
+
+Each class to be tested requires two files for unit testing:
+1. C++ unit test source file (e.g. test_NetworkInterface.cpp)
+2. CMake module file for unit test suite (unittest.cmake)
+
+#### Create files manually
+1. Create a directory for unit test files in `UNITTESTS/rtos/Semaphore`.
+2. Create a CMake module file `UNITTESTS/rtos/Semaphore/unittest.cmake` with the following content:
+```
+set(TEST_SUITE_NAME "rtos_Semaphore")
+
+set(unittest-sources ${unittest-sources}
+	stubs/mbed_assert.c
+	../rtos/Semaphore.cpp
+)
+
+set(unittest-test-sources ${unittest-test-sources}
+	rtos/Semaphore/test_Semaphore.cpp
+)
+```
+3. Create a test source file `UNITTESTS/rtos/Semaphore/test_Semaphore.cpp` with the following content:
+```
+#include  "gtest/gtest.h"
+#include  "rtos/Semaphore.h"
+
+static osStatus_t retval = osOK;
+static uint32_t count = 0;
+
+// Test stubs
+osStatus_t osSemaphoreAcquire (osSemaphoreId_t semaphore_id, uint32_t timeout) { return retval; }
+osStatus_t osSemaphoreDelete (osSemaphoreId_t semaphore_id)                    { return retval; }
+osStatus_t osSemaphoreRelease (osSemaphoreId_t semaphore_id)                   { return retval; }
+uint32_t osSemaphoreGetCount (osSemaphoreId_t semaphore_id)                    { return count; }
+osSemaphoreId_t osSemaphoreNew (uint32_t max_count, uint32_t initial_count, const osSemaphoreAttr_t *attr)
+{
+    return (void*)&count; // Just a dymmy reference
+}
+
+class  TestSemaphore : public  testing::Test
+{
+protected:
+	rtos::Semaphore *sem;
+
+	virtual  void  SetUp()
+	{
+		sem = new rtos::Semaphore();
+	}
+
+	virtual  void  TearDown()
+	{
+		delete sem;
+	}
+};
+
+TEST_F(TestSemaphore, constructor)
+{
+    EXPECT_TRUE(sem);
+}
+```
+
```

Comment:
Could you add a section about the CMake variable defined; the one that can be overridden/extended and the one that can must be defined in the _test project_ ?

Reply:
Added sections for CMake variables used for configuring test suites and the project itself.